src/dune_rules/virtual_rules.ml

```diff
@@ -74,59 +74,58 @@ let impl sctx ~(lib : Library.t) ~scope =
          ]
      | Some vlib ->
        let info = Lib.info vlib in
-       let virtual_ =
-         match Lib_info.virtual_ info with
-         | Some v -> v
-         | None ->
-           User_error.raise
-             ~loc:lib.buildable.loc
-             [ Pp.textf
-                 "Library %s isn't virtual and cannot be implemented"
-                 (Lib_name.to_string implements)
-             ]
-       in
-       let+ vlib_modules, vlib_foreign_objects =
-         let foreign_objects = Lib_info.foreign_objects info in
-         match virtual_, foreign_objects with
-         | External _, Local | Local, External _ -> assert false
-         | External modules, External fa -> Memo.return (modules, fa)
-         | Local, Local ->
-           let name = Lib.name vlib in
-           let vlib = Lib.Local.of_lib_exn vlib in
-           let* dir_contents =
-             let info = Lib.Local.info vlib in
-             let dir = Lib_info.src_dir info in
-             Dir_contents.get sctx ~dir
-           in
-           let* ocaml = Context.ocaml (Super_context.context sctx) in
-           let* modules =
-             let db = Scope.libs scope in
-             let* preprocess =
-               (* TODO wrong, this should be delayed *)
-               Preprocess.Per_module.with_instrumentation
-                 lib.buildable.preprocess
-                 ~instrumentation_backend:(Lib.DB.instrumentation_backend db)
-               |> Resolve.Memo.read_memo
-             in
-             Dir_contents.ocaml dir_contents
-             >>= Ml_sources.modules
-                   ~libs:db
-                   ~for_:(Library (Lib_info.lib_id info |> Lib_id.to_local_exn))
-             >>=
-             let pp_spec =
-               Staged.unstage (Pp_spec.pped_modules_map preprocess ocaml.version)
-             in
-             Modules.map_user_written ~f:(fun m -> Memo.return (pp_spec m))
-           in
-           let+ foreign_objects =
-             Dir_contents.foreign_sources dir_contents
-             >>| Foreign_sources.for_lib ~name
-             >>| (let ext_obj = ocaml.lib_config.ext_obj in
-                  let dir = Obj_dir.obj_dir (Lib.Local.obj_dir vlib) in
-                  Foreign.Sources.object_files ~ext_obj ~dir)
-             >>| List.map ~f:Path.build
-           in
-           modules, foreign_objects
-       in
-       Some (Vimpl.make ~impl:lib ~vlib ~vlib_modules ~vlib_foreign_objects))
+       (match Lib_info.virtual_ info with
```

Comment:
You could remove this pattern match and just turn this into a pre-condition.

Reply:
done